feat: add config for optional parameters in a chat message

NJordan72 opened the pull request:
# Description

Currently only role/content properties are passed into the chat template (with the exception of some hardcoded optional params that are model specific).  This PR creates a configuration property to add additional optional parameters.

## Motivation and Context

- Chat templates are not restricted to using role/content properties and we should allow users, especially those creating their own custom chat templates to pass in anything via the message.
- OPEN QUESTION: Why restrict the properties passed into the chat template in the first place?  Couldn't we just pass the entire message in and if the template doesn't use a particular property they just become no-ops?

## How has this been tested?

- Added unit test

NJordan72 committed:
feat: add config for optional parameters in a chat message

NJordan72 committed:
chore: empty commit

NJordan72 committed:
chore: cleanup

winglian commented on src/axolotl/prompt_strategies/chat_template.py:
```diff
@@ -44,6 +45,13 @@ def __init__(
             }
 
+        # Default optional fields, kept for compatibility
+        self._default_optional_fields = ["tool_calls", "name", "tool_call_id"]
```
Can we include the comments/descriptions on these fields from `get_conversation_thread`?

winglian commented on src/axolotl/prompt_strategies/chat_template.py:
```diff
@@ -44,6 +45,13 @@ def __init__(
+        self.optional_message_fields = self._default_optional_fields + (
+            optional_message_fields or []
+        )
+
```
Might be helpful to use some sort of set operation or other uniqueness check in case a user duplicates in their configuration a key from the default optional fields

winglian commented on tests/prompt_strategies/test_chat_templates_advanced.py:
```diff
@@ -912,5 +912,84 @@ def verify_labels(labels_span, should_train, context_message):
+        input_ids = res["input_ids"]
+
+        LOG.info(f"Labels: {labels}")
+        LOG.info(f"Input IDs: {input_ids}")
```
```suggestion
        LOG.debug(f"Labels: {labels}")
        LOG.debug(f"Input IDs: {input_ids}")
```
log probably not needed in the tests, but can convert these to debug if we want to keep them.

winglian commented on tests/prompt_strategies/test_chat_templates_advanced.py:
```diff
@@ -912,5 +912,84 @@ def verify_labels(labels_span, should_train, context_message):
+        # Verify both optional fields are in the tokenized output
+        decoded_output = tokenizer.decode(input_ids)
+
+        LOG.info(f"Decoded output: {decoded_output}")
```
```suggestion
        LOG.debug(f"Decoded output: {decoded_output}")
```

winglian reviewed (commented):
A few minor nits, but overall looks like a good improvement to chat templates.

NJordan72 committed:
chore: fix nits and add light docs

NJordan72 commented:
> A few minor nits, but overall looks like a good improvement to chat templates.

fixed.

NanoCode012 commented on docs/dataset-formats/conversation.qmd:
```diff
@@ -46,6 +46,9 @@ datasets:
     message_field_content: value
+    optional_message_fields:
+
+
```
I don't think we need this part here to keep it short and simple
```suggestion
```

NanoCode012 commented on tests/prompt_strategies/test_chat_templates_advanced.py:
```diff
@@ -912,5 +912,84 @@ def verify_labels(labels_span, should_train, context_message):
+        turn_labels = labels[start_idx:end_idx]
+        assert not all(
+            label == IGNORE_TOKEN_ID for label in turn_labels
+        ), "Expected assistant message content to be trained on"
```
```suggestion
        assert all(
            label != IGNORE_TOKEN_ID for label in turn_labels
        ), "Expected assistant message content to be trained on"
```

I think this is the condition we want?

Edit: This test may fail until you set `train_on_eos: turn`.

NanoCode012 commented on docs/config.qmd:
```diff
@@ -137,6 +137,8 @@ datasets:
     # Key for content in each message (default:  "content")
     message_field_content: content
+    # Fields that will be passed to the chat template if present in the message. (Optional[List[str]], default: None)
+    optional_message_fields:
```
Should we say we include `tool_calls`, `name`... by default based on your implementation?

NJordan72 replied:
Happy to go either way.  Right now we include them as hardcoded, but I'd be happy to refactor it to just make them the default.

NJordan72 replied:
I guess the downside of making it the default is that if someone wants to add an additional field, they -- at that point -- have to remember to add back the defaults as well.

It is why my slight preference is just to pass the entire message into the chat template and let the template itself decide what it cares about.

NJordan72 replied:
The other option... Would be to parse out all of the arguments from the jinja template and pass in any/all of those.  Jinja provides a pretty easy way to extract the variables from the AST (per documentation, I haven't tried it yet).

We could even warn if the template is expecting something it isn't being passed which might be a nice quality of life addition.

winglian replied:
Also, should this be a `List[str]` (current)

or a `List[Dict[str, str]]` which would allow us to map alternate field names to the field in the jinja template?

The latter more closely matches the meaning of the other args like `message_field_content`, `message_field_role`

NanoCode012 reviewed (commented):
Thanks for the PR. 

> Why restrict the properties passed into the chat template in the first place?

The initial code was written when it was mainly role/content. I expanded it recently for tool_calling datasets, so this is a welcome change.

>  Couldn't we just pass the entire message in and if the template doesn't use a particular property they just become no-ops?

This could be an alternative, but I think the current solution works fine?

NJordan72 committed:
docs: update docs/dataset-formats/conversation.qmd

Co-authored-by: NanoCode012 <kevinvong@rocketmail.com>

NJordan72 committed:
feat: configurable message mappings, jinja template analyzer

NJordan72 committed:
chore: handle bradley terry

NJordan72 commented:
I just updated this to be a little more thoughtful / robust...

- I removed the `optional_message_fields` option and replaced it with a `message_property_mappings` option.
- `message_property_mappings` is basically a more generalized version of the `message_field_role` and `message_field_content` configuration options.  Those continue to work and are just sucked into the `message_property_mappings` if provided.  The idea is you can map any input property to any message property using this configuration.
- I then had to update `chat_template.py` to use the new mappings configuration property to transform the message.
- I added a `JinjaTemplateAnalyzer` that introspects the jinja template to figure out which properties a template is expecting (this is generalized, but it can also look at the properties specifically for the messages, which is what we care about here.
- We then remove any properties from the transformed message that the template does not need/expect.
- I've taken out special handling of `tool_call_id`, etc. as it is no longer needed an is handled by the introspection of the template itself
- I've added test coverage for all of the new stuff and all of the old tests continue to pass

NJordan72 committed:
docs: update docs

winglian commented on tests/prompt_strategies/test_chat_templates.py:
```diff
@@ -74,8 +76,10 @@ def test_llama3(self, llama3_tokenizer, assistant_dataset):
+                message_property_mappings={
+                    "from": "role",
+                    "value": "content",
+                },
```
I think it might be better to invert the property mappings so that the key is `role` and `content`. Because we always know we need to expect those keys, and the mapping is the field in the dataset that we should expect to find them.

Just for the sake of completeness and being able to verify the previous tests in CI and ensure backwards compatibility, It might be good to keep these tests with `message_field_role`/`message_field_content` and add a new test suite for message_property_mappings, even if only for a single model architecture.

NJordan72 replied:
on it!

NJordan72 replied:
These specific tests are hard to keep as they were, because the signature of the function is no longer expecting `message_field_role/message_field_content`

Instead we are shuttling `message_field_role/message_field_content` into `message_property_mappings` as part of the config validation.

I have added a test to the model to make sure they are indeed making it into the mappings.

winglian commented:
Here's the stack trace locally of the broken test in CI currently

```
  File "/Users/wing/Projects/ml/axolotl/src/axolotl/prompt_strategies/chat_template.py", line 275, in tokenize_prompt
    input_ids = self.prompter.build_prompt(turns)
                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wing/Projects/ml/axolotl/src/axolotl/prompt_strategies/chat_template.py", line 84, in build_prompt
    return self.tokenizer.apply_chat_template(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wing/.pyenv/versions/3.11.10/envs/axolotl-3.11/lib/python3.11/site-packages/transformers/tokenization_utils_base.py", line 1683, in apply_chat_template
    rendered_chat = compiled_template.render(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/wing/.pyenv/versions/3.11.10/envs/axolotl-3.11/lib/python3.11/site-packages/jinja2/environment.py", line 1304, in render
    self.environment.handle_exception()
  File "/Users/wing/.pyenv/versions/3.11.10/envs/axolotl-3.11/lib/python3.11/site-packages/jinja2/environment.py", line 939, in handle_exception
    raise rewrite_traceback_stack(source=source)
  File "<template>", line 1, in top-level template code
jinja2.exceptions.UndefinedError: 'dict object' has no attribute 'role'
```

and the dataset configuration that is working on main but is broken in the tests

```yaml
chat_template: llama3
datasets:
  - path: mlabonne/FineTome-100k
    type: chat_template
    split: train[:20%]
    field_messages: conversations
    message_field_role: from
    message_field_content: value
```

NJordan72 committed:
refactor: change order of mappings, improve message transform

NJordan72 commented:
Still looking into the e2e failure (if this push didn't fix it).

NJordan72 committed:
refactor: make chat awware of property mappings

NJordan72 committed:
chore: remove .python-version

NJordan72 committed:
chore: revert change

NJordan72 committed:
chore: add dataset validation to tests where appropriate

NJordan72 committed:
chore: add dataset validation to tests where appropriate

NJordan72 committed:
chore: clean up handling of ds_cfg

NJordan72 commented:
I tracked down the problem with the failing e2e test.  I'm sure there are some further nits, feedback which I'm happy to incorporate.  Some other thoughts below that explain some of the changes and then maybe a broader topic that we can take off of this thread if there is any interest.

--

I will say that the one challenge I had is there seems to be a lack of consistency as to if the config that is getting passed around to all of the various classes/functions is supposed to be the `Dict` representation of the Pydantic model or the model itself.

Many of the tests are obviously just constructing a `Dict` that represents the config which obviously makes sense as to why they would want to pass that around, but it feels like it would be nicer if the Pydantic model was the lingua franca across the entire project.

It feels like the config is the "special sauce" in some ways for the project.  Instead of having to write 1000 lines of boilerplate code to get something running, just craft a fairly simple (I'm probably being a little generous here) config file and you're on your way.  I know a lot of the complexity of the project lies in how to deal with that config file, but it feels like the config itself is something of a second class citizen insofar that the `Dict` representation of it seems to be used in a bunch of places and there are bits of the config that seem under documented and somewhat unclear.

Maybe I'm not seeing the forest from the trees here (and Python is certainly not my bailiwick) , but I'd be interested in helping tighten things up around this if there is any interest in it.

NJordan72 committed:
chore: recursively serialize config

winglian commented on src/axolotl/utils/config/models/input/v0_4_1/__init__.py:
```diff
@@ -311,6 +311,9 @@ class KTODataset(BaseModel):
     revision: Optional[str] = None
 
 
+DatasetConfig = Union[SFTDataset, DPODataset, KTODataset]
```
Did you intend to incorporate this into the line far below?

```
datasets: Optional[conlist(Union[SFTDataset, DPODataset, KTODataset], min_length=1)] = None  # type: ignore
```

NJordan72 replied:
yep. thanks.

winglian commented:
@NJordan72 Thanks for your thoughts and insights. We agree that the docs are definitely lacking and we are on a mission this first few months of the year to improve in that area.  The evolution of the project started with a basic DefaultDict with naive validation functions written against it. Only last year did we convert the validation to use pydantic, hence the sort of mismatch of using pydantic up front and then the conversion to a dict after that.  We're definitely open to making the pydantic model the common language through the project, but in places there will still be a mismatch when we're using the lower level dataclasses that we're extending from HF transformers. I'd love your thoughts on how you might approach this.

winglian commented on tests/e2e/patched/test_fa_xentropy.py:
```diff
@@ -76,7 +76,9 @@ def test_lora_packing_fa_cross_entropy(self, temp_dir, gradient_accumulation_ste
             cfg.fp16 = True
 
+        validate_config(cfg)
         normalize_config(cfg)
```
this is a good call to add the validation in this test as it surfaces a new bug with parsing the chat_dataset. It seems we don't do this consistently across the rest of the tests, and seems like we should add this to more of them to make sure the change for the chat_dataset class doesn't affect other logic.

winglian committed:
make sure to use the return value from validate_config

winglian committed:
DefaultDict pickle/unpickle fix

winglian committed:
fix super call for override

NanoCode012 commented on src/axolotl/prompt_strategies/chat_template.py:
```diff
@@ -44,8 +44,10 @@ def __init__(
-        self.message_field_content = message_field_content
+        self._chat_template_msg_variables = self.get_chat_template_msg_variables(
+            chat_template
+        )
```
If I understand correctly, `messages_array_name` refers to the key of the `List[dict]`. Should we change the signature so that `field_messages` is passed to the Prompter as well? To allow passing `messages_array_name=field_messages`?

NJordan72 replied:
Good catch.  Just updated that.

NanoCode012 commented on src/axolotl/utils/config/models/input/v0_4_1/__init__.py:
```diff
@@ -184,6 +186,66 @@ class SFTDataset(BaseModel):
+                data["message_field_role"] or "role"
+            )
+        elif "role" not in data["message_property_mappings"]:
+            data["message_property_mappings"]["role"] = "role"
```
After setting the property, should we drop the old fields? (Same for content below)

NJordan72 replied:
I think I'll leave the property there for now.  My intention is to come back and implement #2271 in a way that gives us much better type safety and validation logic, at which point it should be easier to to

winglian commented on src/axolotl/prompt_strategies/messages/chat.py:
```diff
@@ -51,17 +51,16 @@ def load(tokenizer, cfg, ds_cfg: Optional[Dict[str, Any]] = None):
+    if message_property_mappings and "role" in message_property_mappings:
+        builder_kwargs["message_field_role"] = message_property_mappings["role"]
+    if message_property_mappings and "content" in message_property_mappings:
+        builder_kwargs["message_field_content"] = message_property_mappings["content"]
```
These changes should probably reverted and go into chat_template.py. We can address changing the messages type in a later PR.

NJordan72 replied:
reverted in the most recent revision

NJordan72 commented:
fyi, will crank the changes out today and tomorrow.

NJordan72 committed:
refactor: message fields

NJordan72 committed:
refactor: message fields

NJordan72 committed:
tests: validate config before using

NJordan72 committed:
chore: add config validation to all e2e tests

NJordan72 committed:
chore: add unneeded logging

NJordan72 committed:
chore: add missed config validation

NJordan72 committed:
chore: pass field_messages to prompter

NJordan72 committed:
test: fix borked test

NJordan72 commented:
@winglian @NanoCode012 looks like the remaining failure is unrelated to my changes here

winglian commented on src/axolotl/utils/data/test.sql.s:
did you accidentally commit this?

winglian commented on src/axolotl/utils/config/models/input/v0_4_1/__init__.py:
```diff
@@ -1644,3 +1669,65 @@ def check_torch_compile_auto(cls, data):
+            data["message_field_content"] or "content"
+        )
+    elif "content" not in data["message_property_mappings"]:
+        data["message_property_mappings"]["content"] = "content"
```
Since it seems like we're starting a deprecation of message_field_role and message_field_content by referring to them as legacy, we should probably raise a deprecation warning here. Also, we should update https://github.com/axolotl-ai-cloud/axolotl/blob/main/scripts/chat_datasets.py#L33-L54 to reflect the new preferred schema.

NJordan72 replied:
added the warning and updated that script

NJordan72 committed:
chore: remove uninteded file

NJordan72 committed:
chore: add deprecation warning and update chat_datasets script

NJordan72 committed:
chore: lint

winglian commented:
@NJordan72 I rebased this from one of your earlier commits against latest main, and cherry-picked a few of your later fixes to clean up the commit history on this. I hope I managed to grab all the rest of your recent changes.

NJordan72 commented:
> @NJordan72 I rebased this from one of your earlier commits against latest main, and cherry-picked a few of your later fixes to clean up the commit history on this. I hope I managed to grab all the rest of your recent changes.

lgtm

NanoCode012 commented on src/axolotl/utils/dict.py:
```diff
@@ -13,3 +13,26 @@ def __missing__(self, key):
 
     def __or__(self, other):
         return DictDefault(super().__ror__(other))
+
+    def __setitem__(self, name, value):
+        # workaround for pickle/unpickle issues and __frozen not being available
+        try:
+            isFrozen = hasattr(  # pylint: disable=invalid-name
+                self, "__frozen"
+            ) and object.__getattribute__(self, "__frozen")
+        except AttributeError:
+            isFrozen = False  # pylint: disable=invalid-name
+
+        if isFrozen and name not in super().keys():
+            raise KeyError(name)
+        super(Dict, self).__setitem__(name, value)  # pylint: disable=bad-super-call
+        try:
+            p = object.__getattribute__(self, "__parent")
+            key = object.__getattribute__(self, "__key")
+        except AttributeError:
+            p = None
+            key = None
+        if p is not None:
+            p[key] = self
+            object.__delattr__(self, "__parent")
+            object.__delattr__(self, "__key")
```
What does this section do exactly? Why do we need it?

winglian replied:
I added this, something is happening in the distributed handling where the config get's pickled, but because there is now a DictDefault as a nested dataset configuration, it has to unpickle that and it fails.

NanoCode012 commented on src/axolotl/prompt_strategies/chat_template.py:
```diff
@@ -25,10 +26,10 @@ def __init__(
+        message_property_mappings: Optional[Dict[str, str]] = None,
         message_field_training: Optional[str] = None,
         message_field_training_detail: Optional[str] = None,
+        messages_array_name: str = "messages",
```
I'm of the opinion to remove `_messages` and its utility function from ChatTemplateStrategy and to call it via `self.prompter.messages_array_name`. This way, we won't have to pass the same variable to two places.

This change would also require updating all tests to remove `strategy.messages = ...` and within the `load()` within this file.

NJordan72 replied:
Removed.

NanoCode012 commented on src/axolotl/prompt_strategies/chat_template.py:
```diff
@@ -212,6 +225,10 @@ def __init__(
         self.train_on_eos = train_on_eos
         self.images = "images"
 
+        LOG.info(
```
```suggestion
        LOG.debug(
```

Lower the priority log level here.

NanoCode012 commented on src/axolotl/utils/config/models/input/v0_4_1/__init__.py:
```diff
@@ -893,6 +902,14 @@ def deprecate_sharegpt_datasets(cls, datasets):
+        if ds_configs:
+            return [ds_config.model_dump() for ds_config in ds_configs]
+        return None
+
```
I did some local tests and noticed a lot of keys which were previously omitted, now exist within the cfg as `key: None`. This messes with some settings like: `"messages_array_name": dataset_config.get("field_messages", "messages")`, returning `messages_array_name: None` (which causes `get_conversation_thread` to fail at  `prompt[self.messages]`).

This doesn't cause error in our tests because we haven't called `strategy.messages` or manually set it ourselves.

Removing this section fixed it for me.

NJordan72 replied:
Added a exlude_none to the model_dump

NanoCode012 commented on src/axolotl/prompt_strategies/chat_template.py:
```diff
@@ -424,61 +441,82 @@ def find_turn(self, turns: list[dict], turn_idx: int):
+
             turn = {
-                "role": self.prompter.roles[message[self.prompter.message_field_role]],
+                **transformed_message,
```
Would we need to worry about any nested copying here? Should we set `turn = transformed_message` then set the other two fields after?

NanoCode012 commented on src/axolotl/utils/config/models/input/v0_4_1/__init__.py:
```diff
@@ -214,26 +225,40 @@ def check_chat_template_config(cls, data):
+    field_messages: Optional[str] = None
+    field_label: Optional[bool] = None
+    prompt_format: Optional[str] = None
+    completion_format: Optional[str] = None
```
I almost missed this as we didn't have test to cover it.

This would change the nested layer for UserDefined for DPO/KTO. What is the motivation behind this change?

NJordan72 replied:
Reverted.

NJordan72 committed:
feat: update axolotlinputconfig and test_models

- add configdict import in axolotl/utils/config/models/input/v0_4_1/__init__.py
- remove unnecessary line breaks in sftdataset, dpodataset, ktodataset, stepwisesuperviseddataset classes
- update model_dump method in axolotlinputconfig to exclude none values
- correct typo in test_models.py comment

NJordan72 committed:
feat: simplify dpodataset and ktodataset classes in config models

removed several optional fields from dpodataset and ktodataset classes in axolotl/utils/config/models/input/v0_4_1. this simplifies the configuration subsets for these datasets.

NJordan72 committed:
feat: improve readability and structure in dataset configuration models

this commit enhances the readability and structure of the dataset configuration models in the `axolotl/utils/config/models/input/v0_4_1` module. it removes unused `configdict` import and adds line breaks to separate class definitions for better clarity. additionally, a minor documentation fix is included to ensure a newline at the end of the `stepwise_supervised.qmd` file.

NJordan72 committed:
feat: change log level from info to debug in chattemplatestrategy

NJordan72 committed:
feat(prompt_strategies): refactor chattemplateprompter and chattemplatestrategy

- Make `chat_template` a required parameter in `ChatTemplatePrompter` constructor
- Add default value for `message_property_mappings` in `ChatTemplatePrompter` constructor
- Add `messages_array_name` property to `ChatTemplatePrompter`
- Change `processor` type to Optional in `ChatTemplatePrompter`
- Add TypeError check for `processor` in `ChatTemplatePrompter.build_prompt`
- Remove `_messages` property from `ChatTemplateStrategy`
- Make `prompter` a required parameter and add type hint in `ChatTemplateStrategy` constructor
- Remove `messages` getter and setter from `ChatTemplateStrategy`
- Use `prompter.messages_array_name` in `ChatTemplateStrategy.get_conversation_thread`
- Remove condition to set `messages` field in `load` function

NJordan72 committed:
feat(tests/utils): ignore type check in load_model call in test_models.py

NJordan72 committed:
feat: improve type handling and test structure in chat templates

- Add return type hint for `get_chat_template` function in `chat_templates.py`
- Remove unnecessary assignment of `strategy.messages` in several test cases
- Add `messages_array_name` parameter to various test configurations in `test_chat_templates.py` and `test_chat_templates_advanced.py`
- Remove redundant `strategy.messages` assignment in `test_chat_templates_advanced.py`

NJordan72 committed:
feat(axolotl): enhance chat strategy with datasetconfig support

This commit introduces support for DatasetConfig in the ChatTemplateStrategy. It also refines the strategy loader to handle different types of ds_cfg inputs and improves the clarity of the code by formatting and reordering. The key changes include:

- Importing Union from typing and BaseModel from pydantic.
- Adding DatasetConfig as an optional type for ds_cfg in StrategyLoader.
- Adjusting the handling of ds_cfg in StrategyLoader to account for BaseModel instances.
- Refactoring the prompter_params and strategy_params for better readability.
- Changing the reference from prompt[self.messages] to prompt[self.prompter.messages_array_name] in the is_prompt_batched method.

NJordan72 committed:
feat: update message handling in btchattemplatestrategy

* Replace `self.messages` with direct string references to "chosen_messages" and "rejected_messages"
* Append system, user, and assistant content directly to "chosen_messages" and "rejected_messages"
* Add a new attribute "messages_array_name" to the `load` function parameters
* Remove the conditional attribute assignment for "field_messages" in the `load` function

NJordan72 committed:
feat: add config validation in test_kd.py

- Import `validate_config` from `axolotl.utils.config`
- Validate the configuration in `test_llama_kd` and another function in `TestKnowledgeDistillation` class

NJordan72 committed:
feat: enhance config validation and capabilities handling

* Import `EnvCapabilities` and `GPUCapabilities` from `axolotl.utils.config.models.internals`
* Update `validate_config` function to create `KTODataset` and `SFTDataset` instances using `dict(ds_cfg)`
* Replace `capabilities` and `env_capabilities` with instances of `GPUCapabilities` and `EnvCapabilities` respectively in `AxolotlConfigWCapabilities` model dump

NJordan72 committed:
feat: update config validation in axolotl utils

- Remove import of `EnvCapabilities` and `GPUCapabilities` from `axolotl.utils.config.models.internals`
- Update `validate_config` function to use `capabilities` and `env_capabilities` directly instead of creating new instances of `GPUCapabilities` and `EnvCapabilities`

NJordan72 committed:
feat: refactor strategyloader in chat_template.py

- Extracted the creation of strategy parameters into a separate function, `_get_strategy_params(cfg, dataset_config)`
- Created a new function, `_get_strategy_cls()`, to obtain the strategy class
- Replaced `ChatTemplateStrategy` with `strategy_cls` for strategy instantiation

NJordan72 committed:
trigger CI

NJordan72 committed:
chore: revert dataset config changes for kto/dpo

NJordan72 committed:
subject: refactor: rename 'messages_array_name' to 'field_messages'

Body:
- Renamed 'messages_array_name' to 'field_messages' in 'ChatTemplatePrompter' class and its usages in 'chat_template.py'
- Updated 'load' function in 'bradley_terry/chat_template.py' to reflect the change
- Adjusted 'get_chat_template_msg_variables' and 'get_message_vars' methods in 'jinja_template_analyzer.py' to use the new variable name
- Modified 'StrategyLoader' in 'chat_template.py' to use 'field_messages'
- Updated tests in 'test_chat_templates.py' and 'test_chat_templates_advanced.py' to use 'field_messages' instead of 'messages_array_name'

NJordan72 committed:
feat: refactor prompt strategies and update config models

* Remove redundant 'return None' in `axolotl/prompt_strategies/__init__.py`
* Simplify message handling in `axolotl/prompt_strategies/bradley_terry/chat_template.py` by using a single 'messages' list instead of separate 'chosen_messages' and 'rejected_messages' lists
* Update default 'message_property_mappings' in `axolotl/prompt_strategies/bradley_terry/chat_template.py`
* Add 'field_messages' field to `axolotl/utils/config/models/input/v0_4_1/__init__.py` configuration model

NJordan72 commented:
@winglian @NanoCode012 sorry for the spam here, but I rebased this with main and fixed what i think are the remaining problems.

winglian commented:
thanks, I rebased it again, @NanoCode012 let's get this merged this week

NanoCode012 committed:
chore: remove unused input

NanoCode012 committed:
chore: remove redundant type ignore

NanoCode012 committed:
fix: remove old configs and update examples

NanoCode012 committed:
fix: type check

NanoCode012 committed:
fix: remove loading old config in ChatMessage

NanoCode012 committed:
fix: update faq with potential new undefinederror

NanoCode012 committed:
fix: add debug if property mapped is not found

NanoCode012 committed:
chore: improve explanation for unmapped properties

NanoCode012 committed:
fix: update docs with new config

NanoCode012 committed:
chore: add note for deprecation config and del old config from dict

NanoCode012 commented on src/axolotl/utils/config/models/input/v0_4_1/__init__.py:
```diff
@@ -683,17 +702,15 @@ class Config:
 
-    datasets: Optional[conlist(Union[SFTDataset, DPODataset, KTODataset, StepwiseSupervisedDataset], min_length=1)] = None  # type: ignore
-    test_datasets: Optional[conlist(Union[SFTDataset, DPODataset, KTODataset, StepwiseSupervisedDataset], min_length=1)] = None  # type: ignore
+    datasets: Optional[conlist(DatasetConfig, min_length=1)] = None  # type: ignore
+    test_datasets: Optional[conlist(DatasetConfig, min_length=1)] = None  # type: ignore
```
This will have merge conflict with the #2309

NJordan72 replied:
If you want to merge this one first, I can address the conflict in #2309.

NanoCode012 replied:
Sounds good!

NanoCode012 reviewed (approved):
Thank you the PR. Sorry it took some time for the review as it had many files.

I rebased it again and removed ref to the configs in docs/yml.

We might need to address type conflict with PR #2309 mentioned.

winglian commented on tests/test_validation_dataset.py:
```diff
@@ -302,3 +302,22 @@ def test_dataset_sharegpt_deprecation(self, minimal_cfg):
+            | {
+                "datasets": [
+                    {
+                        "path": "mhenrichsen/alpaca_2k_test",
+                        "type": "alpaca",
+                        "message_property_mappings": {
+                            "role": "role",
+                            "content": "content",
+                        },
```
Is this just to validate that the config validates? Not actually verifying against the dataset, right?

NJordan72 replied:
correct

NJordan72 committed:
Merge branch 'axolotl-ai-cloud:main' into feat/optional-chat-template-fields
